x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLService.java

```diff
@@ -483,33 +496,77 @@ X509ExtendedTrustManager wrapWithDiagnostics(X509ExtendedTrustManager trustManag
                     .limit(2) // we only need to distinguishing between 0/1/many
                     .map(Entry::getKey)
                     .collect(Collectors.toUnmodifiableList());
+                final String name;
                 switch (names.size()) {
                     case 0:
-                        return "(unknown)";
+                        name = "(unknown)";
+                        break;
                     case 1:
-                        return names.get(0);
+                        name = names.get(0);
+                        break;
                     default:
-                        return "(shared)";
+                        name = "(shared)";
+                        break;
                 }
+                return name + " (with trust configuration: " + configuration.getTrustConfig() + ")";
             };
             trustManager = new DiagnosticTrustManager(trustManager, contextName, diagnosticLogger::warn);
         }
         return trustManager;
     }
 
-    public static Map<String, SSLConfiguration> getSSLConfigurations(Settings settings) {
+    public static Map<String, SslConfiguration> getSSLConfigurations(Environment env) {
+        return getSSLConfigurations(env, env.settings());
+    }
+
+    private static Map<String, SslConfiguration> getSSLConfigurations(Environment env, Settings settings) {
         final Map<String, Settings> sslSettingsMap = getSSLSettingsMap(settings);
-        final Map<String, SSLConfiguration> sslConfigurationMap = new HashMap<>(sslSettingsMap.size());
+        final Map<String, SslConfiguration> sslConfigurationMap = new HashMap<>(sslSettingsMap.size());
         sslSettingsMap.forEach((key, sslSettings) -> {
             if (key.endsWith(".")) {
                 // Drop trailing '.' so that any exception messages are consistent
                 key = key.substring(0, key.length() - 1);
             }
-            sslConfigurationMap.put(key, new SSLConfiguration(sslSettings));
+            try {
+                sslConfigurationMap.put(key, SslSettingsLoader.load(sslSettings, null, env, getKeyStoreFilter(key)));
+            } catch (SslConfigException e) {
+                throw new ElasticsearchSecurityException("failed to load SSL configuration [{}] - {}", e, key, e.getMessage());
+            }
         });
         return Collections.unmodifiableMap(sslConfigurationMap);
     }
 
+    private static Function<KeyStore, KeyStore> getKeyStoreFilter(String sslContext) {
```

Comment:
This implements the logic that @BigPandaToo & I discussed.
It filters out CAs cert/keys from any keystore where there are both CAs and non-CA keypairs.

Because there was a version of this already in X-Pack, I needed to replicate it or the enrollment tests would start to fail. Rather than replicate logic that we knew wasn't quite what we wanted, I implemented something that was closer to our desired end state.

Reply:
Should we add a note in breaking changes for this ? I know this is rather improbable but there can be keystores in use right now that have 2 PrivateKeyEntry objects, one with CA key/cert and one with a non-CA , where the CA entry is used by the JDK code because its alias is returned first. This change will make it so the non-CA entry will be used and that will break clients connecting to it.

Reply:
I think it's worth adding a deprecation log.
It's a scenario we can detect (using the same logic here) and issue a deprecation(style) warning.

That's probably more useful than an item breaking changes which everyone needs to read, but no one will care about.  

I opened https://github.com/elastic/elasticsearch/issues/76710